### PR TITLE
Convert libcudf resource parameters to rmm::device_async_resource_ref

### PR DESCRIPTION
**Pull request by harrism**
## Description
Closes https://github.com/rapidsai/cudf/issues/15498

For reviewers:
Almost all of the thousands of changes are simple textual replace of `rmm::mr::device_memory_resource *` with `rmm::device_async_resource_ref`. 

I think the only substantial changes that are different are in `contiguous_split.cu` (which was assigning `nullptr` to the MR pointer -- I have changed these cases to use a `std::optional<rmm::device_async_resource_ref>`), and in JNI code.

~I still need to figure out how to build and test the JNI bindings. And figure out necessary Cython changes.~

JNI is passing CI now. Cython required no changes.

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by bdice** [comment]
It looks like there are some unrelated CI changes in the diff. Could you take a peek?

File: java/src/main/native/src/TableJni.cpp [hunk 1]
```diff
@@ -46,6 +46,7 @@
 #include <cudf/utilities/span.hpp>
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/resource_ref.hpp>
```
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:49]: Is this needed, or were other changes meant to be made in this file?

Comment (harrism) [reply java/src/main/native/src/TableJni.cpp:49]: It was added by my script but then I undid the changes made by the script, so I need to remove this.

**Comment by harrism** [comment]
> It looks like there are some unrelated CI changes in the diff. Could you take a peek?

Definitely are. I have no idea how they got them. Any idea how to remove them, or do I have to start over for the third time?

**Comment by bdice** [comment]
@harrism Thanks. I think the problem with the diff was that commit https://github.com/rapidsai/cudf/pull/15507/commits/7f49c979c392a800a58d9e4c8feb6d8f4512ca5f wasn't a normal merge commit that included the history (which is then excluded from the diff, because the branches have common history).

**Comment by harrism** [comment]
I have added some guidance for reviewers to the description.

File: java/src/main/native/src/RmmJni.cpp [hunk 2]
```diff
@@ -572,9 +573,10 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_Rmm_allocInternal(JNIEnv *env, jclas
                                                               jlong stream) {
   try {
     cudf::jni::auto_set_device(env);
-    rmm::mr::device_memory_resource *mr = rmm::mr::get_current_device_resource();
+    rmm::device_async_resource_ref mr = rmm::mr::get_current_device_resource();
     auto c_stream = rmm::cuda_stream_view(reinterpret_cast<cudaStream_t>(stream));
-    void *ret = mr->allocate(size, c_stream);
+    // TODO need to pass alignment here
+    void *ret = mr.allocate_async(size, c_stream);
```
[the remark this answers is not in the document]
Comment (harrism) [reply java/src/main/native/src/RmmJni.cpp]: Done.

File: java/src/main/native/src/TableJni.cpp (hunk 1 above)
Comment (harrism) [reply java/src/main/native/src/TableJni.cpp]: Removed.

File: cpp/include/cudf/detail/search.hpp [hunk 3]
```diff
@@ -47,24 +48,24 @@ std::unique_ptr<column> upper_bound(table_view const& haystack,
...
- * @copydoc cudf::contains(column_view const&, scalar const&, rmm::mr::device_memory_resource*)
+ * @copydoc cudf::contains(column_view const&, scalar const&, rmm::device_async_resource_ref )
  *
  * @param stream CUDA stream used for device memory operations and kernel launches.
  */
 bool contains(column_view const& haystack, scalar const& needle, rmm::cuda_stream_view stream);
 
 /**
- * @copydoc cudf::contains(column_view const&, column_view const&, rmm::mr::device_memory_resource*)
+ * @copydoc cudf::contains(column_view const&, column_view const&, rmm::device_async_resource_ref )
```
Comment (bdice) [inline cpp/include/cudf/detail/search.hpp:61]: A lot of these docstrings have an extra space at the end after `rmm::device_async_resource_ref`. Can we fix that? e.g.
```suggestion
 * @copydoc cudf::contains(column_view const&, column_view const&, rmm::device_async_resource_ref)
```

**Review by jlowe — APPROVED** [review]

**Review by bdice — APPROVED** [review]
I read the first ~200 files and spot-checked the rest. I have one comment on spacing in docstrings. Otherwise, this looks good to me.

(I also looked at `contiguous_split.cu` specifically as requested.)

File: cpp/include/cudf/detail/search.hpp (hunk 3 above)
Comment (harrism) [reply cpp/include/cudf/detail/search.hpp]: Done.

**Comment by harrism** [comment]
Thank you for the reviews. @bdice thoughts on whether we should proceed with merging this out of sync with other repos? I kind of want to wait until I have a bit more confident progress on other repos, especially RAFT, which is proving kind of tricky.

**Comment by bdice** [comment]
> Thank you for the reviews. @bdice thoughts on whether we should proceed with merging this out of sync with other repos? I kind of want to wait until I have a bit more confident progress on other repos, especially RAFT, which is proving kind of tricky.

I'm somewhat inclined to merge sooner, given the scope of this change and its potential for conflicts. It would also be nice to have a large RAPIDS repository using the new code paths, to ensure that it's working as intended. However, if you want to wait, I'm okay with that.

**Comment by harrism** [comment]
I got the RAFT changes working. Whether they like them is a different matter. But I agree, this PR looks pretty clean and safe, it is reviewed, and at risk of a lot of conflicts, so let's go ahead and merge

**Comment by harrism** [comment]
Hmmm, doesn't seem to be obeying my merge command. There is a lock symbol by @mythrocks, I wonder if his review is required for some reason?

**Pull request merged**